File: src/main/java/mainPackage/CodeTextArea.java
package mainPackage;


import java.io.FileOutputStream;
import java.io.PrintWriter;
import java.util.Collection;
import java.util.Collections;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javafx.application.Application;
import javafx.scene.Scene;
import javafx.scene.control.ContextMenu;
import javafx.scene.control.MenuItem;
import javafx.scene.layout.StackPane;
import javafx.stage.Stage;

import org.fxmisc.flowless.VirtualizedScrollPane;
import org.fxmisc.richtext.CodeArea;
import org.fxmisc.richtext.LineNumberFactory;
//import org.fxmisc.richtext.demo.JavaKeywordsAsync;
import org.fxmisc.richtext.model.StyleSpans;
import org.fxmisc.richtext.model.StyleSpansBuilder;
import windows.CustomTextArea;

public class CodeTextArea {

    private static final String[] KEYWORDS = new String[] {
            "abstract", "assert", "boolean", "break", "byte",
            "case", "catch", "char", "class", "const",
            "continue", "default", "do", "double", "else",
            "enum", "extends", "final", "finally", "float",
            "for", "goto", "if", "implements", "import",
            "instanceof", "int", "interface", "long", "native",
            "new", "package", "private", "protected", "public",
            "return", "short", "static", "strictfp", "super",
            "switch", "synchronized", "this", "throw", "throws",
            "transient", "try", "void", "volatile", "while"
    };

    private static final String KEYWORD_PATTERN = "\\b(" + String.join("|", KEYWORDS) + ")\\b";
    private static final String PAREN_PATTERN = "\\(|\\)";
    private static final String BRACE_PATTERN = "\\{|\\}";
    private static final String BRACKET_PATTERN = "\\[|\\]";
    private static final String SEMICOLON_PATTERN = "\\;";
    private static final String STRING_PATTERN = "\"([^\"\\\\]|\\\\.)*\"";
    private static final String COMMENT_PATTERN = "//[^\n]*" + "|" + "/\\*(.|\\R)*?\\*/";

    private static final Pattern PATTERN = Pattern.compile(
            "(?<KEYWORD>" + KEYWORD_PATTERN + ")"
                    + "|(?<PAREN>" + PAREN_PATTERN + ")"
                    + "|(?<BRACE>" + BRACE_PATTERN + ")"
                    + "|(?<BRACKET>" + BRACKET_PATTERN + ")"
                    + "|(?<SEMICOLON>" + SEMICOLON_PATTERN + ")"
                    + "|(?<STRING>" + STRING_PATTERN + ")"
                    + "|(?<COMMENT>" + COMMENT_PATTERN + ")"
    );

    private static final String sampleCode = String.join("\n", new String[] {
            "package com.example;",
            "",
            "import java.util.*;",
            "",
            "public class Foo extends Bar implements Baz {",
            "",
            "    /*",
            "     * multi-line comment",
            "     */",
            "    public static void main(String[] args) {",
            "        // single-line comment",
            "        for(String arg: args) {",
            "            if(arg.length() != 0)",
            "                System.out.println(arg);",
            "            else",
            "                System.err.println(\"Warning: empty string as argument\");",
            "        }",
            "    }",
            "",
            "}"
    });

    MainController mainController;

    public MainController getMainController() {
        return mainController;
    }

    public void setMainController(MainController mainController) {
        this.mainController = mainController;

    }

    public CodeTextArea(MainController mainController) {
        this.mainController = mainController;
        codeArea = new org.fxmisc.richtext.CodeArea();
        codeArea.setParagraphGraphicFactory(LineNumberFactory.get(codeArea));
//
        codeArea.richChanges()
                .filter(ch -> !ch.getInserted().equals(ch.getRemoved())) // XXX
                .subscribe(change -> {
                    codeArea.setStyleSpans(0, computeHighlighting(codeArea.getText()));
                });

        //mainController.mainSplitPane.getScene().getStylesheets().add(JavaKeywordsAsync.class.getResource("java-keywords.css").toExternalForm());


        setText(sampleCode);


        ContextMenu cm = new ContextMenu();

        MenuItem menuItem = new MenuItem("Save");

        cm.getItems().add(menuItem);

        codeArea.setWrapText(true);



        codeArea.setContextMenu(cm);


        menuItem.setOnAction(ex->{
            String path = mainController.pathLabelContent.getText();

            try (PrintWriter pw = new PrintWriter(new FileOutputStream(path))) {

                pw.print(this.codeArea.getText());

                System.err.println("___________" + Thread.currentThread().getStackTrace()[1].getClassName()+ "____Line:" + Thread.currentThread().getStackTrace()[1].getLineNumber() +
                "___ " + codeArea.getText());

                mainController.loadingTask.updateMessage("Saved " + path + "!");
            } catch (Exception e) {
                e.printStackTrace();
                CommonUtilities.showErrorAlert("Could not save " + path);
            }
        });

    }

    public CodeTextArea() {


    }

    org.fxmisc.richtext.CodeArea codeArea;


    void setText(String text){

        codeArea.replaceText(0,0,text);

        System.err.println("___________" + Thread.currentThread().getStackTrace()[1].getClassName()+ "____Line:" + Thread.currentThread().getStackTrace()[1].getLineNumber() +
                "___ " +  codeArea.getText());

    }

    VirtualizedScrollPane getCodeArea(){



        return new VirtualizedScrollPane<>(codeArea);


    }




    private static StyleSpans<Collection<String>> computeHighlighting(String text) {
        Matcher matcher = PATTERN.matcher(text);
        int lastKwEnd = 0;
        StyleSpansBuilder<Collection<String>> spansBuilder
                = new StyleSpansBuilder<>();
        while(matcher.find()) {
            String styleClass =
                    matcher.group("KEYWORD") != null ? "keyword" :
                            matcher.group("PAREN") != null ? "paren" :
                                    matcher.group("BRACE") != null ? "brace" :
                                            matcher.group("BRACKET") != null ? "bracket" :
                                                    matcher.group("SEMICOLON") != null ? "semicolon" :
                                                            matcher.group("STRING") != null ? "string" :
                                                                    matcher.group("COMMENT") != null ? "comment" :
                                                                            null; /* never happens */ assert styleClass != null;
            spansBuilder.add(Collections.emptyList(), matcher.start() - lastKwEnd);
            spansBuilder.add(Collections.singleton(styleClass), matcher.end() - matcher.start());
            lastKwEnd = matcher.end();
        }
        spansBuilder.add(Collections.emptyList(), text.length() - lastKwEnd);
        return spansBuilder.create();
    }
}

File: src/main/kotlin/mainPackage/CodeTextArea.kt
package mainPackage

import javafx.scene.control.ContextMenu
import javafx.scene.control.MenuItem
import org.fxmisc.flowless.VirtualizedScrollPane
import org.fxmisc.richtext.LineNumberFactory
import org.fxmisc.richtext.model.StyleSpans
import org.fxmisc.richtext.model.StyleSpansBuilder
import java.io.FileOutputStream
import java.io.PrintWriter
import java.util.Collections
import java.util.regex.Pattern

class CodeTextArea {
    var mainController: MainController? = null
    var codeArea: org.fxmisc.richtext.CodeArea

    constructor(mainController: MainController) {
        this.mainController = mainController
        codeArea = org.fxmisc.richtext.CodeArea()
        codeArea.paragraphGraphicFactory = LineNumberFactory.get(codeArea)
        codeArea.richChanges()
            .filter { ch -> ch.inserted != ch.removed }
            .subscribe { codeArea.setStyleSpans(0, computeHighlighting(codeArea.text)) }
        setText(SAMPLE_CODE)
        val cm = ContextMenu()
        val menuItem = MenuItem("Save")
        cm.items.add(menuItem)
        codeArea.isWrapText = true
        codeArea.contextMenu = cm
        menuItem.setOnAction {
            val path = mainController.pathLabelContent.text
            try {
                PrintWriter(FileOutputStream(path)).use { pw ->
                    pw.print(codeArea.text)
                    MainController.loadingTask.updateMessage("Saved $path!")
                }
            } catch (e: Exception) {
                e.printStackTrace()
                CommonUtilities.showErrorAlert("Could not save $path")
            }
        }
    }

    constructor()  {
        codeArea = org.fxmisc.richtext.CodeArea()
    }

    fun setText(text: String) { codeArea.replaceText(0, 0, text) }
    fun getCodeArea(): VirtualizedScrollPane<org.fxmisc.richtext.CodeArea> = VirtualizedScrollPane(codeArea)

    companion object {
        private val KEYWORDS = arrayOf(
            "abstract", "assert", "boolean", "break", "byte", "case", "catch", "char", "class", "const",
            "continue", "default", "do", "double", "else", "enum", "extends", "final", "finally", "float",
            "for", "goto", "if", "implements", "import", "instanceof", "int", "interface", "long", "native",
            "new", "package", "private", "protected", "public", "return", "short", "static", "strictfp", "super",
            "switch", "synchronized", "this", "throw", "throws", "transient", "try", "void", "volatile", "while"
        )
        private val KEYWORD_PATTERN = "\\b(${KEYWORDS.joinToString("|")})\\b"
        private val PAREN_PATTERN = "\\(|\\)"
        private val BRACE_PATTERN = "\\{|\\}"
        private val BRACKET_PATTERN = "\\[|\\]"
        private val SEMICOLON_PATTERN = "\\;"
        private val STRING_PATTERN = "\"([^\"\\\\]|\\\\.)*\""
        private val COMMENT_PATTERN = "//[^\n]*|/\\*(.|\\R)*?\\*/"
        private val PATTERN = Pattern.compile(
            "(?<KEYWORD>$KEYWORD_PATTERN)|(?<PAREN>$PAREN_PATTERN)|(?<BRACE>$BRACE_PATTERN)|(?<BRACKET>$BRACKET_PATTERN)|(?<SEMICOLON>$SEMICOLON_PATTERN)|(?<STRING>$STRING_PATTERN)|(?<COMMENT>$COMMENT_PATTERN)"
        )
        private val SAMPLE_CODE = listOf(
            "package com.example;", "", "import java.util.*;", "",
            "public class Foo extends Bar implements Baz {", "",
            "    /*", "     * multi-line comment", "     */",
            "    public static void main(String[] args) {",
            "        // single-line comment",
            "        for(String arg: args) {",
            "            if(arg.length() != 0)", "                System.out.println(arg);",
            "            else", "                System.err.println(\"Warning: empty string as argument\");",
            "        }", "    }", "", "}"
        ).joinToString("\n")

        private fun computeHighlighting(text: String): StyleSpans<Collection<String>> {
            val matcher = PATTERN.matcher(text)
            var lastKwEnd = 0
            val spansBuilder = StyleSpansBuilder<Collection<String>>()
            while (matcher.find()) {
                val styleClass = when {
                    matcher.group("KEYWORD") != null -> "keyword"
                    matcher.group("PAREN") != null -> "paren"
                    matcher.group("BRACE") != null -> "brace"
                    matcher.group("BRACKET") != null -> "bracket"
                    matcher.group("SEMICOLON") != null -> "semicolon"
                    matcher.group("STRING") != null -> "string"
                    matcher.group("COMMENT") != null -> "comment"
                    else -> null
                }!!
                spansBuilder.add(Collections.emptyList(), matcher.start() - lastKwEnd)
                spansBuilder.add(Collections.singleton(styleClass), matcher.end() - matcher.start())
                lastKwEnd = matcher.end()
            }
            spansBuilder.add(Collections.emptyList(), text.length - lastKwEnd)
            return spansBuilder.create()
        }
    }
}
